crates/static-analysis-kernel/src/analysis/ddsa_lib/js/flow/graph.rs

```diff
@@ -0,0 +1,799 @@
+                            (&src.0, &target.0, attr_string(&edge.attributes));
+                        (format!("{src}->{target}{attrs}"), stmt)
+                    }
+                    _ => unreachable!(),
+                })
+                .collect::<HashMap<_, _>>()
+        }
+
+        let self_stmts = normalize_stmts(&self.graph);
+        let other_stmts = normalize_stmts(&other.graph);
+
+        let mut diff = Vec::<Stmt>::new();
+        for (key, &value) in &self_stmts {
+            if !other_stmts.contains_key(key) {
+                diff.push(value.clone());
+            }
+        }
+        diff
+    }
+
+    /// Returns this graph in DOT form. Note that the nodes and edges are displayed in construction-order,
+    /// which does not hold semantic meaning. Two [`FlowDigraph`] can have a different `to_dot` output
+    /// but represent the same graph.
+    pub fn to_dot(&self) -> String {
+        use graphviz_rust::printer::DotPrinter;
+        self.graph.print(&mut Default::default())
+    }
+
+    /// Returns `true` if the graph has no nodes and no edges.
+    pub fn is_empty(&self) -> bool {
+        Self::stmts(&self.graph).is_empty()
+    }
+
+    /// Returns the [`Stmt`](dot_structures::Stmt)s in a [`dot_structures::Graph`].
+    fn stmts(graph: &dot_structures::Graph) -> &[dot_structures::Stmt] {
+        let dot_structures::Graph::DiGraph { stmts, .. } = graph else {
+            unreachable!();
+        };
+        stmts
+    }
+
+    /// Constructs a [`Self`] from the provided `inner`, performing minimal validation.
+    #[cfg(test)]
+    fn from_inner(inner: dot_structures::Graph) -> Self {
+        let dot_structures::Graph::DiGraph { strict, .. } = inner else {
+            panic!("graph must be a digraph");
+        };
+        assert!(strict, "graph must be strict");
+        Self { graph: inner }
+    }
+}
+
+impl PartialEq for FlowDigraph {
+    fn eq(&self, other: &Self) -> bool {
+        self.difference(other).is_empty() && other.difference(self).is_empty()
+    }
+}
+
+/// A [`v8::Number`] used to store a target [`NodeId`](ddsa_lib::common::NodeId) and [`EdgeKind`].
+///
+/// Internally, this is laid out as:
+/// ```text
+///            27 bits           4 bits
+/// |---------------------------|----|
+///            NodeId            Kind
+/// ```
+struct PackedEdge(u32);
+
+impl PackedEdge {
+    /// The number of bits used to represent the uint form of this enum.
+    const KIND_BITS: u32 = 4;
+    /// A bitmask to extract an `EdgeType` from a packed edge.
+    const KIND_BIT_MASK: u32 = (1 << Self::KIND_BITS) - 1;
+
+    /// Returns the target of this edge.
+    pub fn target(&self) -> ddsa_lib::common::NodeId {
+        self.0 >> Self::KIND_BITS
+    }
+
+    /// The type of edge this is.
+    pub fn kind(&self) -> EdgeKind {
+        EdgeKind::try_from_id((self.0 & Self::KIND_BIT_MASK) as usize)
+            .expect("js should serialize PackedEdge correctly")
+    }
+}
```

Comment:
Why this? This adds more complexity than anything else.

Why not a struct with a nodeid and a kind?